### PR TITLE
NIAD-1096: Add PPRF (or REC) participant to participant2 when mapping encounters to ehr compositions

### PR DESCRIPTION
**Pull request by andrewwa-kainos**
AC10 - add Participant2 element to ehrComposition

GIVEN Encounter is translated to ehrComposition
AND Encounter.participant of type “PPRF” is translated to Agent
AND if there’s no Encounter.participant of type “PPRF” then the one of type “REC” is used
THEN Participant2 element is added to ehrComposition
AND Participant2/agentRef is referencing Agent

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java:63]: Make sure not to override the change in https://github.com/nhsconnect/integration-adaptor-gp2gp/pull/156/files to make this mandatory.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java:66]: should be ::get to fail if the reference doesn't exist

Comment (andrewwa-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java:66]: Did I understand the spreadsheet wrong?
> Populate the participant from the participant of type PPRF and if there is no PPRF participant make the REC participant the PPRF, All other additional participants are ignored

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java:65]: #156 adds a test for the missing recorder reference; I'd like to wait until that is merged to main (and then this) before updating the tests for the missing performer. Will just make the work simpler.

**Review by msolecka — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/parameters/EncounterTemplateParameters.java
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/parameters/EncounterTemplateParameters.java:19]: author is not used in this PR but it does not matter as soon as #156 will be merged first.

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapperTest.java
Comment (msolecka) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapperTest.java:123]:
```suggestion
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java
Comment (andrewwa-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterMapper.java]: Test now added

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/IdMapper.java
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/IdMapper.java:40]: Pinched from #158

**Review by msolecka — APPROVED** [review]

File: service/src/test/resources/ehr/mapper/encounter/example-encounter-resource-15.json
Comment (z-uddin) [inline service/src/test/resources/ehr/mapper/encounter/example-encounter-resource-15.json:39]: I may have misunderstood the test but
Believe this file is used as INPUT_JSON_WITHOUT_PERFORMER_PARTICIPANT in EncounterMapperTest
But this resource has participant of type performer?

Comment (andrewwa-kainos) [reply service/src/test/resources/ehr/mapper/encounter/example-encounter-resource-15.json]: Well, that was incredibly dumb of me. Tests passed because we're mocking everything to have the same ID. Now fixed

**Review by z-uddin — APPROVED** [review]

File: service/src/main/resources/templates/ehr_encounter_to_ehr_composition_template.mustache
Comment (andrewwa-kainos) [inline service/src/main/resources/templates/ehr_encounter_to_ehr_composition_template.mustache:28]: Will change from triple braces to double

**Review by ElliotB-96 — DISMISSED** [review]

**Review by ElliotB-96 — APPROVED** [review]

**Review by msolecka — APPROVED** [review]

**Review by z-uddin — APPROVED** [review]

**Pull request merged**
